src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -413,24 +414,121 @@ private static string EmitIndexOfAnyValues(char[] asciiChars, Dictionary<string,
+            }
+
+            string fieldName = EmitIndexOfAnyValues(asciiChars.ToArray(), requiredHelpers);
+            string? helperName = set switch
+            {
+                RegexCharClass.DigitClass => "IndexOfAnyDigit",
+                RegexCharClass.ControlClass => "IndexOfAnyControl",
+                RegexCharClass.LetterClass => "IndexOfAnyLetter",
+                RegexCharClass.LetterOrDigitClass => "IndexOfAnyLetterOrDigit",
+                RegexCharClass.LowerClass => "IndexOfAnyLower",
+                RegexCharClass.NumberClass => "IndexOfAnyNumber",
+                RegexCharClass.PunctuationClass => "IndexOfAnyPunctuation",
+                RegexCharClass.SeparatorClass => "IndexOfAnySeparator",
+                RegexCharClass.SpaceClass => "IndexOfAnyWhiteSpace",
+                RegexCharClass.SymbolClass => "IndexOfAnySymbol",
+                RegexCharClass.UpperClass => "IndexOfAnyUpper",
+                RegexCharClass.WordClass => "IndexOfAnyWordChar",
+                _ => null,
+            };
+            if (helperName is null)
+            {
+                using (SHA256 sha = SHA256.Create())
+                {
+                    helperName = $"IndexOfNonAsciiOrAny_{BitConverter.ToString(sha.ComputeHash(Encoding.UTF8.GetBytes(set))).Replace("-", "")}";
+                }
+            }
+
+            if (!requiredHelpers.ContainsKey(helperName))
+            {
+                var additionalDeclarations = new HashSet<string>();
+
+                var lines = new List<string>()
+                {
+                    $"/// <summary>Finds the next index of any character that matches {EscapeXmlComment(DescribeSet(set))}.</summary>",
+                    $"[MethodImpl(MethodImplOptions.AggressiveInlining)]",
+                    $"internal static int {helperName}(this ReadOnlySpan<char> span)",
+                    $"{{",
+                    $"    int i = span.IndexOfAnyExcept({fieldName});",
+                    $"    if ((uint)i < (uint)span.Length)",
+                    $"    {{",
+                    $"        if (span[i] <= 0x7f)",
+                    $"        {{",
+                    $"            return i;",
+                    $"        }}",
+                    $"",
+                    $"        do",
+                    $"        {{",
+                    $"            if ({MatchCharacterClass("span[i]", set, negate: false, additionalDeclarations, requiredHelpers)})",
+                    $"            {{",
+                    $"                return i;",
+                    $"            }}",
+                    $"            i++;",
+                    $"        }}",
+                    $"        while ((uint)i < (uint)span.Length);",
+                    $"    }}",
+                    $"",
+                    $"    return -1;",
+                    $"}}",
+                };
+
+                // The below insert positions must be kept in sync with the lines above.
```

Comment:
NIT: I know that the above declaration looks better than a bunch of `lines.Add()`, but I wonder if due to these below additional things we should instead use lines.Add and do the checks on the right place for conditionally adding the code or not. I just feel that it would be easy to break the way this is setup right now, as not everyone always reads all of the comments :P.

Reply:
Done